psi4/driver/procrouting/scf_proc/diis.py

```diff
@@ -0,0 +1,200 @@
+from enum import Enum
+from psi4 import core
+from psi4.driver import psifiles as psif
+
+import numpy as np
+
+class RemovalPolicy(Enum):
+    LargestError = 1
+    OldestAdded = 2
+
+class StoragePolicy(Enum):
+    InCore = 1
+    OnDisk = 2
+
+def axpy(y, alpha, x):
+    if isinstance(y, (core.Matrix, core.Vector)):
+        y.axpy(alpha, x)
+    elif isinstance(y, (core.dpdbuf4, core.dpdfile2)):
+        y.axpy_matrix(x, alpha)
+    else:
+        raise Exception("Unrecognized object type for DIIS.")
+
+def template_helper(template, *args):
+    if template:
+        raise Exception("Template already set.")
+
+    for arg in args:
+        if isinstance(arg, core.Vector):
+            template.append([arg.dimpi()])
+        elif isinstance(arg, (core.Matrix, core.dpdfile2, core.dpdbuf4)):
+            template.append([arg.rowdim(), arg.coldim()])
+        else:
+            raise Exception("Unrecognized object type for DIIS.")
+
+class DIIS:
+
+    def __init__(self, max_vecs: int, name: str, removal_policy = RemovalPolicy.LargestError, storage_policy = StoragePolicy.OnDisk):
+        assert isinstance(removal_policy, RemovalPolicy)
+        assert isinstance(storage_policy, StoragePolicy)
+        self.max_vecs = max_vecs
+        self.name = name
+        self.removal_policy = removal_policy
+        self.storage_policy = storage_policy
+        self.R_template = []
+        self.T_template = []
+        self.reset_subspace()
+        if self.storage_policy == StoragePolicy.OnDisk:
+            psio = core.IO.shared_object()
+            if not psio.open_check(psif.PSIF_LIBDIIS):
+                psio.open(psif.PSIF_LIBDIIS, 1) # 1 = PSIO_OPEN_OLD
+
+    def __del__(self):
+        psio = core.IO.shared_object()
+        if psio.open_check(psif.PSIF_LIBDIIS):
+            psio.close(psif.PSIF_LIBDIIS, 1) # 1 = KEEP
```

Comment:
I would prefer the file to be closed only if this object opened it.  Specifically, if line 50 (the `psio.open` call) did not execute then line 55 should not execute.

Reply:
This is another [feature addition](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libdiis/diismanager.cc#L545-L547), but one I can do easily enough.

Reply:
I wonder if a context manager could be useful here. link is one that I wrote, not one that's particularly relevant https://github.com/MolSSI/QCEngine/blob/master/qcengine/util.py#L558-L559

Reply:
I don't think so. Because file I/O is expensive, we want it to be closed solely on object delete. That outlives any particular function call. The current version seems best, unless I am massively misunderstanding context managers.